core/state/statedb.go

```diff
@@ -289,6 +297,9 @@ func (s *StateDB) TxIndex() int {
 func (s *StateDB) GetCode(addr common.Address) []byte {
 	stateObject := s.getStateObject(addr)
 	if stateObject != nil {
+		if s.witness != nil {
+			s.witness.AddCode(stateObject.Code(s.db))
```

Comment:
Does this witness include everything (code, storage trie nodes, account trie nodes)?

Reply:
Yes (although not needed in our case) it also includes block headers for the `blockhash()` function.

Reply:
Where are the block headers added? I don't see `AddBlockHash` called anywhere.